### PR TITLE
Parallel Model Implementation

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Technical Description

This PR implements the parallel execution model in Dawn as described by the [SIR Execution Model](https://github.com/MeteoSwiss-APN/dawn/wiki/SIR-Execution-Model).

### Resolves / Enhances

Implements Epic #836 
Resolves Issues #882, #916, #961

### Testing

See `dawn_gtclang_clang-gridtools_PR/650/`.

### CODE REVIEW

File: dawn/src/dawn/Optimizer/PassOptions.inc [hunk 1]
```diff
@@ -72,13 +76,13 @@ OPT(bool, ReportPassStageSplit, false, "report-pass-stage-split", "",
...
-    "Report on all field renamings during the field-versioning pass", "", false, true)
+    "Report on all field renaming during the field-versioning pass", "", false, true)
 OPT(bool, ReportPassTemporaryMerger, false, "report-pass-temporary-merger", "",
     "Report which temporaries will be merged during the temporary merger pass. The groups of temporaries, "
-    "which will be merged, are sorted in alphapetical order.", "", false, true)
+    "which will be merged, are sorted in alphabetical order.", "", false, true)
 OPT(bool, ReportPassTemporaryType, false, "report-pass-temporary-type", "",
     "Report which variables and temporary fields are promoted/demoted", "", false, true)
-OPT(bool, ReportPassStageReodering, false, "report-pass-stage-reordering", "",
+OPT(bool, ReportPassStageReordering, false, "report-pass-stage-reordering", "",
```
Comment (jdahm) [inline dawn/src/dawn/Optimizer/PassOptions.inc:85]: 👍

File: dawn/src/dawn/Validator/MultiStageChecker.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  iir::StencilInstantiation* instantiation_;
+  int maxHaloPoints_;
+
+public:
+  MultiStageChecker(iir::StencilInstantiation* instantiation, const int maxHaloPoints = 3);
+
+  void run();
+};
+
+} // namespace dawn
```
Comment (jdahm) [inline dawn/src/dawn/Validator/MultiStageChecker.h:39]: Does MultiStageChecker need to have internal state?

File: dawn/test/integration-test/from-sir/TestComputeMaximumExtent.cpp [hunk 3]
```diff
@@ -12,35 +12,36 @@
...
-#include "dawn/Serialization/SIRSerializer.h"
-#include "dawn/Support/FileSystem.h"
-#include <fstream>
-#include <gtest/gtest.h>
-#include <streambuf>
-#include <string>
+#include "TestFromSIR.h"
 
-using namespace dawn;
+namespace dawn {
```
Comment (jdahm) [inline dawn/test/integration-test/from-sir/TestComputeMaximumExtent.cpp:17]: I think our other tests are in the anonymous namespace. Is there a reason this was added inside the dawn namespace?

File: gtclang/test/integration-test/Regression/globals_dependencies_issue401_ref.0.iir [hunk 4]
```diff
@@ -597,5 +597,5 @@
   ],
   "boundaryConditions": []
  },
- "filename": "/Users/johannd/Code/dawn/gtclang/test/integration-test/Regression/globals_dependencies_issue401.cpp"
+ "filename": "/home/eddied/Work/dawn/gtclang/test/integration-test/Regression/globals_dependencies_issue401.cpp"
```
Comment (jdahm) [inline gtclang/test/integration-test/Regression/globals_dependencies_issue401_ref.0.iir:600]: ;)

Comment (eddie-c-davis) [reply gtclang/test/integration-test/Regression/globals_dependencies_issue401_ref.0.iir]: Changed...

File: dawn/test/integration-test/from-sir/TestComputeMaximumExtent.cpp (hunk 3 above)
Comment (eddie-c-davis) [reply dawn/test/integration-test/from-sir/TestComputeMaximumExtent.cpp]: No particular reason, no. Changed.

File: dawn/src/dawn/Validator/MultiStageChecker.h (hunk 2 above)
Comment (eddie-c-davis) [reply dawn/src/dawn/Validator/MultiStageChecker.h]: No, this class started life as an independent pass. State variables have been removed.

**Pull request merged**